packages/next/src/client/dev/hot-middleware-client.ts

```diff
@@ -1,18 +1,37 @@
+import type {
+  NextRouter,
+  PrivateRouteInfo,
+} from '../../shared/lib/router/router'
 import connect from '../components/react-dev-overlay/pages/hot-reloader-client'
 import { sendMessage } from '../components/react-dev-overlay/pages/websocket'
 
+// Define a local type for the window.next object
+interface NextWindow {
+  next?: {
+    router?: NextRouter & {
+      components: { [pathname: string]: PrivateRouteInfo }
+    }
+  }
+  __nextDevClientId?: string
+  location: Location
+}
+
+declare const window: NextWindow
+
 let reloading = false
 
 export default (mode: 'webpack' | 'turbopack') => {
   const devClient = connect(mode)
 
   devClient.subscribeToHmrEvent((obj: any) => {
     if (reloading) return
-    // if we're on an error/404 page, we can't reliably tell if the newly added/removed page
-    // matches the current path. In that case, assume any added/removed entries should trigger a reload of the current page
+
+    // Retrieve the router if it's available
+    const router = window.next?.router
+
+    // Determine if we're on an error page or the router is not initialized
     const isOnErrorPage =
-      window.next.router.pathname === '/404' ||
-      window.next.router.pathname === '/_error'
+      !router || router.pathname === '/404' || router.pathname === '/_error'
```

Comment:
@gurkerl83 Probably missing optional chaining here

```suggestion
      !router || router?.pathname === '/404' || router?.pathname === '/_error'
```

Reply:
Where optional chaining makes sense and where it does not.

1. Correctly uses optional chaining to safely access router, preventing errors if window.next is undefined.

```window.next?.router```

2. Access of the mandatory field `router.pathname`

The following type is defined for Router

```
export type BaseRouter = {
  route: string
  pathname: string
...
}
```

- The original PR is **safe** without using ?. on router.pathname because !router ensures router is defined before accessing pathname.
- The suggested change adding ?. to router.pathname is **redundant**; once router is confirmed truthy, pathname is guaranteed to exist.

**Keep Original Code:** 
The existing use of optional chaining on window.next?.router is necessary. Adding it to router.pathname is unnecessary and clutters the code.

Thx!